pex/resolver.py

```diff
@@ -248,8 +248,12 @@ def package_iterator(self, resolvable, existing=None):
 
       if self.__cache_ttl:
         packages = self.filter_packages_by_ttl(packages, self.__cache_ttl)
-        if packages:
-          return packages
+        # Return both cached packages and the pypi packages. The edge case is if a inexact
+        # resolvable finds a cached version, it needs to return that cached package AND all pypi
+        # packages because the cached version might not match another resolvable which has a tighter
+        # bound or an exact version.
+        return packages + super(CachingResolver, self).package_iterator(resolvable,
+                                                                        existing=existing)
 
     return super(CachingResolver, self).package_iterator(resolvable, existing=existing)
```

Comment:
well, you could go a step further and guard the initialization of `packages` altogether by checking for the conditions in which it'd be used - was sort of imagining something along the lines of:

```
  def package_iterator(self, resolvable, existing=None):
    packages = []
    if self.__cache_ttl or resolvable.exact:
      iterator = Iterator(fetchers=[Fetcher([self.__cache])])
      packages = self.filter_packages_by_interpreter(
        resolvable.compatible(iterator),
        self._interpreter,
        self._platform
      )

      if resolvable.exact and packages:
        return packages

      packages = self.filter_packages_by_ttl(packages, self.__cache_ttl)

    return itertools.chain(
      packages,
      super(CachingResolver, self).package_iterator(resolvable, existing=existing)
    )
```

but I'm probably just splitting hairs at this point - fine with any of the above.